### DISCUSSION TITLE
WebAssembly build with NativeAOT-LLVM

### DISCUSSION THREAD

**Discussion by yowl** [AvaloniaUI/Avalonia#14888] [General]
Hi, has anyone attempted to build the Wasm head with NativeAOT-LLVM?  I did this for Uno as an experiment some time ago and it was quite a bit of effort, but NativeAOT-LLVM and Mono are more closely aligned these days so it might not be so hard.

**Comment by maxkatz6** [comment]
No, I don't think anybody tried, but there was an interest.
It will likely involve some wasm backend changes as well. 

From what we need to be working:
1. JS interop like `JSImport` with callbacks and `JSHost.ImportAsync`. I don't think we have any `JSExport`. But it might be needed as a replacement for [`Task Main` call](https://github.com/AvaloniaUI/Avalonia/blob/9646164708adb1a90f54b8a461b543ad82bef980/samples/ControlCatalog.Browser/Program.cs#L14-L25), which is required to initialize Avalonia backend. https://github.com/AvaloniaUI/Avalonia/tree/master/src/Browser/Avalonia.Browser/Interop
2. StaticWebAssets support OR WasmExtraFilesToDeploy support - Avalonia needs to include `avalonia.js` and `storage.js` files from the nuget package via props/targets - https://github.com/AvaloniaUI/Avalonia/tree/master/src/Browser/Avalonia.Browser/build
3. Emscripten GL support - https://github.com/AvaloniaUI/Avalonia/blob/master/src/Browser/Avalonia.Browser/build/Avalonia.Browser.targets#L8C6-L8C31 + https://github.com/AvaloniaUI/Avalonia/blob/9646164708adb1a90f54b8a461b543ad82bef980/src/Browser/Avalonia.Browser/webapp/modules/avalonia/canvas.ts#L16
4. And SkiaSharp working on top of that Emscripten GL. Can be either 2.88 or 3.0 (newer one is experimental).

Preferably to try it with latest nightly builds (master) or [release/11.1](https://github.com/AvaloniaUI/Avalonia/tree/release/11.1) branch.

**Reply by maxkatz6** [reply]
Also, currently browser backend is WebGL only. Software rendering canvas isn't available in the latest master, but if necessary, I can pick it up from my forgotten branches.

**Reply by yowl** [reply]
Thanks, looking at `module?.GL ?? self.AvaloniaGL ?? self.SkiaSharpGL` it can work with just one of these 3?  If so, is it known which is most performant ?

**Reply by yowl** [reply]
Or they are all WebGL I suppose.

**Reply by maxkatz6** [reply]
Yeah, they all are webgl. Just at different times it was different one. Since we dropped support for .NET 7 WASM SDK, and I don't think SkiaSharpGL exists anymore either, only the first one makes the most sense.

**Reply by maxkatz6** [reply]
I.e. before we had such hacks for Emscripten https://github.com/AvaloniaUI/Avalonia/blob/11.0.10/src/Browser/Avalonia.Browser/interop.js#L8. `mergeInto` usage was later removed in master, and replaced with `<EmccExportedRuntimeMethod Include="GL" />`

**Reply by yowl** [reply]
I did have WebGPU working with some manual bindings but haven't tried WebGL, so that will be interesting.  For the JS side we have at least some of `JSImport` working, I will create some tests for the async and callbacks to see what state they are in.

**Reply by maxkatz6** [reply]
Hi!
I am tinkering software renderer support right now, and was wondering if I can try NativeAOT-LLVM running.

Do you know if there is a Hello World app similar to .NET 8 WASM templates? Something like [this](https://github.com/dotnet/runtime/tree/main/src/mono/wasm/templates/templates/browser) but with NativeAOT-LLVM. Specifically with a support of JSImport/JSExport and dotnetjs. 
[This](https://github.com/dotnet/runtimelab/blob/feature/NativeAOT-LLVM/docs/using-nativeaot/compiling.md) doc also doesn't have information on that.

I know it should be possible to run Avalonia now without any JSImport/JSExport, but with some work around dll imports, but I was wondering if we can go easier now.

**Comment by yowl** [comment]
Awesome, I did have skia sharp working with just the software renderer a few years back  https://github.com/yowl/wskia, but hit https://github.com/mono/SkiaSharp/issues/1931  hopefully that problem no longer exists.

**Comment by yowl** [comment]
I see you are having a go from the runtimelab board (https://github.com/dotnet/runtimelab/issues/2434#issuecomment-2071197386)  I've been busy fixing the nuget for NativeAOT-LLVM (merging from upstream) so haven't had a chance to start looking at this.  Let me know if you have any questions with the build or support.

**Comment by yowl** [comment]
The latest packages have better JS support, but there are still some problems getting the template app to start.  For example SkiaSharp uses managed delegates to pass to Skia.  These wont work in AOT and need to be converted to function pointers.  I have started to try that, but building SkiaSharp just for wasm doesnt seem easy.

**Reply by charlesroddie** [reply]
FYI @mattleibow
We are using skiasharp in browser (which currently involves mono/blazor) as a POC and wait on nativeaot for having something release-worthy. I think others will be in the same boat and that skiasharp/nativeaot/wasm is going to be an v impactful combo.

**Reply by maxkatz6** [reply]
I haven't had time recently to do another test on LLVM. But I still planning to do that.
Meanwhile we had some improvements with .NET 9 WASM threading support in Avalonia, making backend a bit more flexible on the way.

**Comment by maxkatz6** [comment]
See https://github.com/AvaloniaUI/Avalonia/issues/16211 for continuation of the discussion
